orttraining/orttraining/core/graph/gradient_builder_base.cc

```diff
@@ -165,5 +188,31 @@ void GradientBuilderBase::HandleBroadcasting(const ArgDef& input_grad,
   }
 }
 
+void GradientBuilderBase::HandleBroadcastingDynamic(const ArgDef& input_grad,
+                                                    const ArgDef& target,
+                                                    const ArgDef& output_grad,
+                                                    const ArgDef& reduce_axes,
+                                                    std::vector<NodeDef>& output) const {
+  ArgDef reduce_grad_arg = IA("ReduceSumTraining_" + input_grad.name + "_for_" + target.name);
+  output.push_back(
+      NodeDef(OpDef{"ReduceSumTraining", kMSDomain, 1},
+              {input_grad,
+               reduce_axes},
+              {reduce_grad_arg},
+              {{"keepdims", ONNX_NAMESPACE::MakeAttribute("keepdims", int64_t(1))},
+               {"noop_with_empty_axes", ONNX_NAMESPACE::MakeAttribute("noop_with_empty_axes", int64_t(1))}}));
+
+  ArgDef target_shape_arg = IA(target.name + "_shape");
+  output.push_back(
+      NodeDef("Shape",
+              {target},
+              {target_shape_arg}));
```

Comment:
This Shape node seems duplicated from the shape nodes from ComputeBroadcastBackwardAxesDynamic()

We should reuse the Shape node output